### ISSUE TITLE
 The case for ad-hoc polymorphism in Gren

### ISSUE THREAD

**Issue by robinheghan** [gren-lang/compiler#37]
The current polymorphism story in Gren is, in my view, incomplete. There is ad-hoc polymorphism in the language, but it's not possible for end-users to create new cases, nor possible to extend the cases that do exist.

To make matters worse, some polymorphic functions (like the `==` operator) are unsafe and can trigger runtime exceptions.

Something must be done.

In order to figure out _what_, we first need to know _why_. The purpose of this document is to attempt to answer the latter question.

## Gren's current use of ad-hoc polymorphism
Gren already supports ad-hoc polymorphism, none of which is exposed for extension outside of the compiler and core packages. I'll list the different cases below, with the intention that any future proposal must address these specific use-cases.

### Polymorphic functions

Functions like  `List.sort` accepts types of the built in `comparable` type class. This type means "it can be any generic type, as long as it has an implementation for the `compare` function". This is the typical use case of ad-hoc polymorphism.

### Polymorphic data structures

Some datastructures, like `Dict` and `Set` requires that some of its properties are `comparable`. This is an extension of the above section.

### Magic functions

Certain functions in Gren isn't actually bound by a type class, but simply accepts any generic type and figures out what the implementation should be at runtime. This includes functions like `==` and `Debug.toString`.  These functions are flexible, and requires very little of the user of such functions, but they also aren't type safe. The usage of `==` can be well typed, but still crash at runtime when passed functions as arguments.

### Effect modules

Effect modules are an interesting, but little known, feature. An effect module is essentially a module which carries state. It can define a command type (for other modules to send it messages) or a subscription type (to periodically send messages).

Effect modules are a hard-coded case in the compiler. It clearly defines an interface, but there's no way to define your own effect module-y things.

## The risks involved
Adding support for user defined ad-hoc polymorphism adds complexity to the language. This is certain. An important thing to consider for any potential solution, considering Gren wants to remain easy-to-learn, is to keep this added complexity to a minimum.

There are several ways to ensure this, but in short:
* Any interface should have names which are easy to understand (to beginners)
* Interfaces should be simple and not do to much
* Don't use an interface when you don't have to (in core libs, which by default sets a precedent for the rest of the community)
* Aim for the simplest solution. Be conservative about features which complicates the type system too much.

Another thing to keep in mind is that ad-hoc polymorphism introduces more abstract code. The consequence of this is less concrete (less helpful) error messages.

In short: done incorrectly, this feature could end up making Gren a worse language than it is today.

That said, I still think we should try.

## Simple solutions (which won't do)
Many have proposed solutions to Gren's current lack of ad-hoc polymorphism, or ways that the current limitations could be softened. Some of these seem reasonable at first glance, but (in my opinion) falls apart on closer inspection. 

Below, I'll comment on the proposed solutions I've seen which have seemed the most reasonable, that doesn't propose adding a significant new feature to the language.

###  Automatic derivation

A common complaint is not being able to extend the builtin `comparable` type class. One proposed solution to this is to change the semantics of the language in such a way, that any type which itself only contains `comparable` types, itself is `comparable`.

This could extend to other potential built in type classes in the future, like `equatable` and `hashable`.

This doesn't require any new syntax to be added to the language, and seems the be a relatively simple feature to add. However, I don't like this idea for the following reasons:

* Hardly discoverable: This feature essentially relies on convention. Which makes it easy for people new to the language to accidentally break an API, or accidentally support a feature. Adding a field could suddenly make a type not `comparable` anymore, or removing a field could make a type `comparable` which you never intended to. This lack of control is not a good thing.
* No opt-out: Does it make sense to compare `Person` types? The type has an `age` and a `fullName` property, both of which are `comparable`, but does it still make sense to compare two values of a `Person`? Maybe it does, and maybe it doesn't, but I think it's clear that it would be nice to have a way to opt out of implementing a contract.
* Inflexible: Say you have a type which represents a row in a SQL database. It makes sense that you can use `==` to compare two rows for equality, but you really only need to check the `id` field. With automatic derivation, `==` (if changed to work with a builtin `equatable` type class) only works if every column in a row is a `comparable` type, but even then it's inefficient to compare all the values, as only rows with the same `id` could be equal. Another problem is that the order of the columns suddenly matter. Maybe you want a field to be order able first by `fieldA` _then_ by `id`, well then you'd need to define your data structure in a way that makes the `id` field come after `fieldA`, and you must then never change this ordering. 

### Double down on functions

Another idea is to simply double down on plain functions. `==` and `comparable` could be replaced by the convention of providing an `equal` and `compare` function for your types, and types like `Dict` could require passing in a `compare` function when creating a new instance.

This could work, and it wouldn't really add much complexity to the language.

My main issue with this is that two different functions still have the same type, so there's no way to express that `Dict String` with `String.compare` is different than `Dict String` with `String.lowercase >> String.compare` in the type system.

In addition, this also comes at the cost of increased memory usage (as polymorphic functions would need to be stored in the type), and it would also require more code in several cases (passing in function references).

## In summary

This issue is all about exposing the problems with todays design, of which there are a few. I believe we need a better story concerning ad-hoc polymorphism in Gren, but we also need to be careful that we don't end up making the language worse as a result. 

Future issues will propose solutions to the problems stated here, and I hope we can have a good discussion regarding those proposals.

Stay tuned.

**Comment by besquared** [comment]
W.r.t. comparable and a few others here is an interesting discussion.

https://github.com/elm/core/pull/929

It's pretty much either type classes or functor modules. Maybe there should be another thread for the type classes approach?

**Comment by robinheghan** [comment]
I'm currently exploring how Gren would look if #81 was implemented. Once that is done, I might do a thread on type classes/protocols.

**Comment by besquared** [comment]
More thoughts on these issues in general from roc lang here https://github.com/roc-lang/roc/issues/2463

**Comment by robinheghan** [comment]
Closing this as #81 is chosen to be the way forward.
